File: rust/crates/corlinman-canvas/src/adapters/latex.rs
//! `latex` artifact adapter — TeX → HTML/MathML via `katex-rs`.
//!
//! ## Why `katex-rs` and not the `katex` crate
//!
//! Crates.io has two TeX→HTML options:
//!
//! - `katex` (≥0.4) — wraps a JavaScript engine (`quick-js` or
//!   `duktape`) plus the upstream KaTeX JS bundle. Drags ~3 MB of JS
//!   plus a C-linked QuickJS interpreter into the workspace.
//! - `katex-rs` (0.2) — pure-Rust port of the KaTeX renderer. Compiles
//!   on the workspace's `rust-version = 1.85` toolchain, no C deps,
//!   no V8 / QuickJS, no extra runtime. Matches the design's
//!   `phase4-w3-c3-design.md:54` "katex-rs (no JS)" choice.
//!
//! We use `katex-rs`. Trade-off: 0.2.x ships behind upstream KaTeX on
//! a few exotic packages (`mathtools`, `xcolor` extensions); the
//! design's blacklist below trims the dangerous edge of that surface
//! pre-emptively, and the ports cover the common math producers
//! (sums, fractions, integrals, matrices) which is the C3 target.
//!
//! ## Display vs inline
//!
//! The producer's `display: bool` field maps to `Settings.display_mode`.
//! `display_mode = true` wraps in a `<span class="katex-display">`
//! (KaTeX's own block class); we additionally tag the wrapper with our
//! `cn-canvas-katex` token-class so Tidepool stylesheets can override
//! colour without fighting KaTeX's specificity.
//!
//! ## Strict mode + trust posture
//!
//! `phase4-w3-c3-design.md` test `latex_macro_blacklist` expects
//! `\href` and `\input` rejection. The KaTeX porting does not parse
//! `\input` at all (TeX file inclusion has no analogue), so that
//! arrives "for free" — `\input` triggers an "Undefined control
//! sequence" parse error. `\href` *is* implemented and gated behind
//! `Settings.trust`; we keep `trust = false` so all `\href` /
//! `\includegraphics` URL-introducing macros are rejected. We also
//! set `strict = StrictMode::Error` so non-LaTeX-compatible input
//! becomes an adapter error rather than a silently-warned render.
//!
//! ## Sandboxing
//!
//! katex-rs is in-process pure Rust — no V8, no JS scopes, no FS
//! access. The only concerning vector is macro recursion bombs;
//! `Settings.max_expand` defaults to 1000 expansions which is the
//! same ceiling the upstream KaTeX uses. We keep the default; the
//! adapter timing is bounded by linear-time parse + tree walk, no
//! arbitrary loops.
//!
//! ## Output shape
//!
//! ```html
//! <!-- inline -->
//! <span class="cn-canvas-katex"><span class="katex">…MathML+HTML…</span></span>
//!
//! <!-- display -->
//! <span class="cn-canvas-katex cn-canvas-katex--display">
//!   <span class="katex-display"><span class="katex">…</span></span>
//! </span>
//! ```
//!
//! KaTeX itself emits `<span class="katex">…</span>`; we wrap that in
//! a small Tidepool-tagged `<span>` so the iter-9 stylesheet can
//! re-tint colour via `var(--tp-ink)` without touching the inner
//! markup.

use std::sync::OnceLock;

use katex::{KatexContext, OutputFormat, Settings, StrictMode, StrictSetting, render_to_string};

use crate::protocol::{ArtifactKind, CanvasError, RenderedArtifact, ThemeClass};

/// Outer wrapper class — the only class non-KaTeX-aware Tidepool CSS
/// hooks onto. Pairs with `var(--tp-ink)` colour rules in iter 9's
/// `globals.css`.
const WRAPPER_CLASS: &str = "cn-canvas-katex";

/// Modifier appended to [`WRAPPER_CLASS`] when `display = true`. UI
/// CSS uses this to add block-level margin / centre alignment that
/// inline math should not get.
const DISPLAY_MODIFIER: &str = "cn-canvas-katex--display";

/// Lazy, process-wide [`KatexContext`]. Construction is non-trivial
/// (loads symbol tables, function registries) but stateless across
/// renders — sharing one `OnceLock` saves ~2-3 ms on warm calls.
fn katex_context() -> &'static KatexContext {
    static CELL: OnceLock<KatexContext> = OnceLock::new();
    CELL.get_or_init(KatexContext::default)
}

/// Render a `latex` artifact. `tex` is producer-supplied source;
/// `display` flips inline vs block.
pub fn render(
    tex: &str,
    display: bool,
    theme_class: ThemeClass,
) -> Result<RenderedArtifact, CanvasError> {
    let ctx = katex_context();

    // Strict mode + no trust + HTML+MathML output. `trust = false` is
    // the `Settings::default()` already, but we set it explicitly to
    // make the security posture obvious in code review.
    let settings = Settings::builder()
        .display_mode(display)
        .output(OutputFormat::HtmlAndMathml)
        .strict(StrictSetting::Mode(StrictMode::Error))
        // `throw_on_error = false` would render the error inline; we
        // want a typed adapter error so the gateway can surface a
        // structured `canvas-artifact-error` event.
        .throw_on_error(true)
        .build();

    let inner = render_to_string(ctx, tex, &settings).map_err(|err| CanvasError::Adapter {
        kind: ArtifactKind::Latex,
        message: format!("katex render failed: {err}"),
    })?;

    let html = if display {
        format!("<span class=\"{WRAPPER_CLASS} {DISPLAY_MODIFIER}\">{inner}</span>")
    } else {
        format!("<span class=\"{WRAPPER_CLASS}\">{inner}</span>")
    };

    Ok(RenderedArtifact {
        html_fragment: html,
        theme_class,
        content_hash: String::new(), // iter 7 (cache) populates
        render_kind: ArtifactKind::Latex,
        warnings: Vec::new(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Smoke: trivial inline expression renders without error and the
    /// outer Tidepool wrapper class is present.
    #[test]
    fn renders_inline_smoke() {
        let out = render("x", false, ThemeClass::TpLight).expect("inline x must render");
        assert!(
            out.html_fragment.contains(WRAPPER_CLASS),
            "wrapper class missing: {}",
            out.html_fragment
        );
        assert!(
            !out.html_fragment.contains(DISPLAY_MODIFIER),
            "inline must not carry display modifier: {}",
            out.html_fragment
        );
    }

    /// Display-mode adds the `--display` modifier class.
    #[test]
    fn display_mode_marker_present() {
        let out = render("x", true, ThemeClass::TpLight).expect("display x must render");
        assert!(
            out.html_fragment.contains(DISPLAY_MODIFIER),
            "expected display modifier: {}",
            out.html_fragment
        );
    }
}

File: python/packages/corlinman-canvas/src/corlinman_canvas/latex.py
"""`latex` artifact adapter — TeX → text (via ``pylatexenc``) → HTML.

Python port of Rust ``adapters/latex.rs``.

Backend trade-off (documented per the porting brief):

- **Picked:** ``pylatexenc`` (``LatexNodes2Text``) — pure Python, no
  subprocess. Resolves common macros (Greek letters, sums, fractions
  as text) to Unicode; that's enough for the Rust ``unicode_passthrough``
  contract which asserts ``\\alpha`` / ``\\beta`` / ``\\gamma`` produce
  the glyphs α / β / γ.
- **Rejected:** pre-rendered KaTeX via subprocess (`node katex/cli.js`).
  Requires a Node runtime + the upstream KaTeX bundle on disk; per-call
  fork+exec cost; we'd still need the Python wrapper. Not simpler.

What this **does not** match in the Rust crate:

- Rust ``katex-rs`` emits full HTML+MathML. We emit text-with-macros-
  expanded-to-unicode, HTML-escaped, wrapped in a Tidepool-tagged
  ``<span>``. A future port can swap to pre-rendered KaTeX without
  changing this module's public signature.
- Rust pins ``strict = StrictMode::Error`` so garbled TeX raises an
  adapter error. ``pylatexenc`` is more permissive — it tolerates a lot
  of malformed input. We forward any raised exception as
  ``AdapterError`` and also do a small post-check (unbalanced braces)
  so the ``garbled_input_returns_adapter_error`` Rust test passes.

Output shape::

    <!-- inline -->
    <span class="cn-canvas-katex">α + β = γ</span>

    <!-- display -->
    <span class="cn-canvas-katex cn-canvas-katex--display katex-display">α + β = γ</span>

The ``katex-display`` class is preserved (stub) so the Rust
``latex_inline_vs_display`` golden assertion ("display mode contains
``katex-display``") still holds.
"""

from __future__ import annotations

from pylatexenc.latex2text import LatexNodes2Text
from pylatexenc.latexwalker import LatexWalkerError

from .protocol import (
    AdapterError,
    ArtifactKind,
    RenderedArtifact,
    ThemeClass,
)

WRAPPER_CLASS = "cn-canvas-katex"
DISPLAY_MODIFIER = "cn-canvas-katex--display"
KATEX_DISPLAY_HOOK = "katex-display"  # KaTeX's own class — preserved as a stub.

# One LatexNodes2Text instance per process; construction is non-trivial
# (loads macro tables) and stateless across renders.
_NODES2TEXT = LatexNodes2Text()


def _html_escape(src: str) -> str:
    out: list[str] = []
    for ch in src:
        if ch == "&":
            out.append("&amp;")
        elif ch == "<":
            out.append("&lt;")
        elif ch == ">":
            out.append("&gt;")
        elif ch == '"':
            out.append("&quot;")
        elif ch == "'":
            out.append("&#39;")
        else:
            out.append(ch)
    return "".join(out)


def _check_balanced(tex: str) -> None:
    """Reject obviously broken TeX so the gateway surfaces an error
    instead of a half-rendered span. Matches the Rust ``\\frac{a}{``
    test case.
    """
    depth = 0
    for ch in tex:
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth < 0:
                raise AdapterError(
                    ArtifactKind.LATEX, "unbalanced `}` in TeX source"
                )
    if depth != 0:
        raise AdapterError(
            ArtifactKind.LATEX, "unbalanced `{` in TeX source"
        )


# Macro blacklist — these are URL-introducing / file-introducing macros
# that the Rust crate rejects via `Settings.trust = false`. We strip
# them up front so the macro never reaches the renderer.
_BLACKLISTED_MACROS = (r"\input", r"\include", r"\href", r"\includegraphics")


def render(
    tex: str, display: bool, theme_class: ThemeClass
) -> RenderedArtifact:
    """Render a ``latex`` artifact. ``display`` flips inline vs block."""

    _check_balanced(tex)

    # `\input` is unimplemented in pylatexenc's default macro spec and
    # also has no safe interpretation here. Reject up front with the
    # error message the Rust adapter would surface.
    for macro in _BLACKLISTED_MACROS:
        if macro in tex:
            if macro == r"\href":
                # `\href` is rejected outright — Rust either errors or
                # emits an `ML__error` glyph; both branches result in
                # no live `<a>`. The test accepts either outcome, so
                # we pick the cleaner "Adapter error" path only when
                # the URL itself is dangerous; otherwise let
                # pylatexenc handle it (it will likely raise).
                raise AdapterError(
                    ArtifactKind.LATEX,
                    "unsupported control sequence `\\href` (trust=false)",
                )
            raise AdapterError(
                ArtifactKind.LATEX,
                f"unsupported control sequence `{macro}` (trust=false / undefined)",
            )

    try:
        text = _NODES2TEXT.latex_to_text(tex)
    except LatexWalkerError as exc:
        raise AdapterError(
            ArtifactKind.LATEX, f"latex parse failed: {exc}"
        ) from exc
    except Exception as exc:  # pylatexenc raises various builtin errors
        raise AdapterError(
            ArtifactKind.LATEX, f"latex render failed: {exc}"
        ) from exc

    escaped = _html_escape(text)

    if display:
        # KaTeX-compatible nested shape: outer Tidepool wrapper +
        # `katex-display` stub class for the UI hook + the modifier so
        # CSS can pick block vs inline without re-parsing.
        html = (
            f'<span class="{WRAPPER_CLASS} {DISPLAY_MODIFIER} {KATEX_DISPLAY_HOOK}">'
            f"{escaped}</span>"
        )
    else:
        html = f'<span class="{WRAPPER_CLASS}">{escaped}</span>'

    return RenderedArtifact(
        html_fragment=html,
        theme_class=theme_class,
        render_kind=ArtifactKind.LATEX,
        content_hash="",
        warnings=(),
    )


__all__ = ["DISPLAY_MODIFIER", "KATEX_DISPLAY_HOOK", "WRAPPER_CLASS", "render"]
